tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+_LOADED_BACKENDS = {}
+
+
+def _initialize_default_backend():
+    """Setup the initial Tensorly backend"""
+    default = 'numpy'
+    backend = os.environ.get('TENSORLY_BACKEND', default)
+
+    if backend not in _KNOWN_BACKENDS:
+        msg = ("TENSORLY_BACKEND should be one of {%s}, got %r. Defaulting to "
+               "'%r'") % (', '.join(map(repr, _KNOWN_BACKENDS)), backend, default)
+        warnings.warn(msg, UserWarning)
+        backend = default
+
+    set_backend(backend, make_default=True)
+
+
+class set_backend(object):
+    """Sets the backend for TensorLy.
+
+    Can be used to set the backend globally, or locally as a contextmanager.
+
+    Parameters
+    ----------
+    backend : {'numpy', 'mxnet', 'pytorch', 'tensorflow', 'cupy'}
+        The name of the backend to use. Default is 'numpy'.
+    make_default : bool, optional
+        If True, the backend will become the default backend for all threads.
+        Note that this only affects threads where the backend hasn't already
+        been explicitly set. If False (default) the backend is only set
+        contextually for the local thread.
+
+    Examples
+    --------
+    Set the backend to numpy globally for this thread:
+
+    >>> import tensorly as tl
+    >>> tl.set_backend('numpy')
+
+    Set the backend to numpy globally for all threads:
+
+    >>> tl.set_backend('numpy', make_default=True)
+
+    Set the backend to use pytorch inside a context:
+
+    >>> with tl.set_backend('pytorch'):  # doctest: +SKIP
+    ...     pass
+    """
+    def __init__(self, backend, make_default=False):
+        if not isinstance(backend, Backend):
+            if backend not in _LOADED_BACKENDS:
+                # load the backend
+                if backend in _KNOWN_BACKENDS:
+                    importlib.import_module('tensorly.backend.%s_backend' % backend)
+                else:
+                    msg = ("Unknown backend %r, known backends are {%s}"
+                           % (backend, ', '.join(map(repr, _KNOWN_BACKENDS))))
+                    raise ValueError(msg)
+            backend = _LOADED_BACKENDS[backend]
+
+        # Set the backend
+        self._old_backend = getattr(_STATE, 'backend', None)
+        self._new_backend = _STATE.backend = backend
+
+        if make_default:
+            global _DEFAULT_BACKEND
+            _DEFAULT_BACKEND = backend
+
+    def __repr__(self):
+        return 'tensorly.set_backend(%r)' % self._new_backend.backend_name
+
+    def __enter__(self):
+        return None
+
+    def __exit__(self, *args):
+        if self._old_backend is None:
+            del _STATE.backend
+        else:
+            _STATE.backend = self._old_backend
+
+
+def get_backend():
+    """Returns the backend currently used
+
+    Returns
+    -------
+    backend_name : str
+        The backend currently in use
+    """
+    return getattr(_STATE, 'backend', _DEFAULT_BACKEND).backend_name
```

Comment:
Seeing these two classes act on essentially the same data, it might make sense to have a BackendManager class? These could be staticmethod? 
Again, this is more for discussion than a hard point :)

Reply:
I assume you mean `set_backend` (a class for implementation reasons) and `get_backend` (a function)? I think these make more sense as top-level definitions - the only state shared is global, and nesting them under a class would overly complicate things.

Reply:
I would be tempted to get rid of the STATE and simply have a single (global) BackendManager storing and handling everything, including registering the backends functions. Thoughts?